universe/vault/credentials.go

```diff
@@ -62,5 +64,39 @@ func (c *Credentials) Login(ctx context.Context, options ...vault.ClientOption)
 		return nil, err
 	}
 
+	go renew(ctx, client, resp.Auth)
+
 	return client, nil
 }
+
+func renew(ctx context.Context, client *vault.Client, auth *vault.ResponseAuth) {
+	lease := time.Duration(auth.LeaseDuration) * time.Second
+	if lease <= 0 {
+		return // token does not expire
+	}
+	if !auth.Renewable {
+		zerolog.Ctx(ctx).Warn().Msgf("vault: non-renewable token expires in %s", lease)
+		return
+	}
+
+	interval := lease - time.Minute*2
+	if interval < 0 {
+		zerolog.Ctx(ctx).Warn().Msgf("vault: not renewing token, lease too short: %s", lease)
+		return
+	}
+
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			if _, err := client.Auth.TokenRenewSelf(ctx, schema.TokenRenewSelfRequest{}); err != nil {
+				// TODO: Let the consumer know, so it could try and reconnect if needed?
+				zerolog.Ctx(ctx).Error().Msgf("vault: failed to renew token: %v", err)
+				return // TODO: retry, with backoff
```

Comment:
constant backoff should be fine here https://github.com/namespacelabs/foundation/blob/7223a6e741b1e532bcb4c44eb36edde7b732e9ae/internal/providers/nscloud/ctl/wait.go#L107-L108

Reply:
Turns out the Vault client library [already handles retries](https://github.com/hashicorp/vault-client-go/blob/cfdcbeda2e34770a045cc9f71ad25db195e97181/client_option.go#L65) so let's not retry here.